src/executor/current_thread.rs

```diff
@@ -0,0 +1,361 @@
+                        `block_with_init`")
+            })
+    }
+}
+
+impl<F> Executor<F> for CurrentThread
+where F: Future<Item = (), Error = ()> + 'static
+{
+    fn execute(&self, future: F) -> Result<(), ExecuteError<F>> {
+        spawn(future, false)
+    }
+}
+
+
+impl<F> Executor<F> for DaemonExecutor
+where F: Future<Item = (), Error = ()> + 'static
+{
+    fn execute(&self, future: F) -> Result<(), ExecuteError<F>> {
+        spawn(future, true)
+    }
+}
+
+fn spawn<F>(future: F, daemon: bool) -> Result<(), ExecuteError<F>>
+where F: Future<Item = (), Error = ()> + 'static,
+{
+    CURRENT.with(|current| {
+        if current.scheduler.get().is_null() {
+            Err(ExecuteError::new(ExecuteErrorKind::Shutdown, future))
+        } else {
+            let spawned = SpawnedFuture {
+                daemon: daemon,
+                inner: Task::new(future),
+            };
+
+            unsafe {
+                (*current.scheduler.get()).push(spawned);
+            }
+
+            Ok(())
+        }
+    })
+}
+
+impl TaskRunner {
+    /// Return a new `TaskRunner`
+    fn new(thread_notify: Arc<ThreadNotify>) -> TaskRunner {
+        TaskRunner {
+            non_daemons: 0,
+            scheduler: scheduler::Scheduler::new(thread_notify),
+        }
+    }
+
+    /// Enter a new `TaskRunner` context
+    fn enter<F, A>(f: F) -> Result<A::Item, A::Error>
+        where F: FnOnce() -> A,
+              A: Future,
+    {
+        // Create a new task runner that will be used for the duration of `f`.
+        ThreadNotify::with_current(|thread_notify| {
+            // The runner has to be created outside of the MY_TASK_RUNNER.with
+            // block.
+            let mut runner = TaskRunner::new(thread_notify.clone());
+
+            CURRENT.with(|current| {
+                // Make sure that another task runner is not set.
+                assert!(current.scheduler.get().is_null());
+
+                // Set the scheduler to the TLS and perform setup work,
+                // returning a future to execute.
+                //
+                // This could possibly spawn other tasks.
+                let future = current.set_scheduler(&mut runner.scheduler, f);
+
+                // Execute the runner
+                runner.finish(thread_notify, current, future)
+            })
+        })
+    }
+
+    fn finish<F: Future>(&mut self,
+                         thread_notify: &Arc<ThreadNotify>,
+                         current: &CurrentRunner,
+                         future: F)
+        -> Result<F::Item, F::Error>
+    {
+        let mut result = None;
+        let mut future = Some(executor::spawn(future));
+
+        while future.is_some() || self.non_daemons > 0 {
+            if current.cancel.get() {
+                // TODO: This probably can be improved
+                current.cancel.set(false);
+
+                debug_assert!(current.scheduler.get().is_null());
```

Comment:
This assert initially caught me by surprise, because I hadn't caught that the scheduler was only temporarily established in `enter`. Can we make this flow more clear?